internal/formatter/formatter.go

```diff
@@ -0,0 +1,149 @@
+package formatter
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/textileio/go-tableland/internal/tableland"
+)
+
+// Output is used to control the output format of a query specified with the "output" query param.
+type Output string
+
+const (
+	// Table returns the query results as a JSON object with columns and rows properties.
+	Table Output = "table"
+	// Objects returns the query results as a JSON array of JSON objects. This is the default.
+	Objects Output = "objects"
+)
+
+var outputsMap = map[string]Output{
+	"table":   Table,
+	"objects": Objects,
+}
+
+// OutputFromString converts a string into an Output.
+func OutputFromString(o string) (Output, bool) {
+	output, ok := outputsMap[o]
+	return output, ok
+}
+
+// FormatConfig is the format configuration used.
+type FormatConfig struct {
+	Output  Output
+	Unwrap  bool
+	Extract bool
+}
+
+// FormatOption controls the behavior of calls to Format.
+type FormatOption func(*FormatConfig)
+
+// WithOutput specifies the output format. Default is Table.
+func WithOutput(output Output) FormatOption {
+	return func(fc *FormatConfig) {
+		fc.Output = output
+	}
+}
+
+// WithUnwrap specifies whether or not to unwrap the returned JSON objects from their surrounding array.
+// Default is false.
+func WithUnwrap(unwrap bool) FormatOption {
+	return func(fc *FormatConfig) {
+		fc.Unwrap = unwrap
+	}
+}
+
+// WithExtract specifies whether or not to extract the JSON object
+// from the single property of the surrounding JSON object.
+// Default is false.
+func WithExtract(extract bool) FormatOption {
+	return func(fc *FormatConfig) {
+		fc.Extract = extract
+	}
+}
+
+// Format transforms the user rows according to the provided configuration, retuning raw json or jsonl bytes.
+func Format(userRows *tableland.TableData, opts ...FormatOption) ([]byte, FormatConfig, error) {
```

Comment:
This function is used in both the user controller and json rpc service. The idea is that those services read data from the core api, `tableland.Tableland` which always returns `tableland.TableData` (formally called `tableland.UserData`), then pass that table data into the formatter with the appropriate `FormatOption`s derived from user input parameters. This function returns `[]byte` that are ready to return to the caller.